streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyInnerJoinIntegrationTest.java

```diff
@@ -0,0 +1,396 @@
+                                                      final Set<KeyValue<Integer, String>> expectedResult) {
+        final String queryableName = joinType + "-ktable-ktable-joinOnForeignKey-query";
+        final ReadOnlyKeyValueStore<Integer, String> myJoinStoreOne = streams.store(queryableName,
+                QueryableStoreTypes.keyValueStore());
+
+        final ReadOnlyKeyValueStore<Integer, String> myJoinStoreTwo = streamsTwo.store(queryableName,
+                QueryableStoreTypes.keyValueStore());
+
+        final ReadOnlyKeyValueStore<Integer, String> myJoinStoreThree = streamsThree.store(queryableName,
+                QueryableStoreTypes.keyValueStore());
+
+        // store only keeps last set of values, not entire stream of value changes
+        final Map<Integer, String> expectedInStore = new HashMap<>();
+        for (KeyValue<Integer, String> expected : expectedResult) {
+            expectedInStore.put(expected.key, expected.value);
+        }
+
+        // depending on partition assignment, the values will be in one of the three stream clients.
+        for (Map.Entry<Integer, String> expected : expectedInStore.entrySet()) {
+            String one = myJoinStoreOne.get(expected.getKey());
+            String two = myJoinStoreTwo.get(expected.getKey());
+            String three = myJoinStoreThree.get(expected.getKey());
+
+            String result;
+            if (one != null)
+                result = one;
+            else if (two != null)
+                result = two;
+            else if (three != null)
+                result = three;
+            else
+                throw new RuntimeException("Cannot find key " + expected.getKey() + " in any of the state stores");
+            assertEquals(expected.getValue(), result);
+        }
+
+        //Merge all the iterators together to ensure that their sum contains the total set of expected elements.
+        final KeyValueIterator<Integer, String> allOne = myJoinStoreOne.all();
+        final KeyValueIterator<Integer, String> allTwo = myJoinStoreTwo.all();
+        final KeyValueIterator<Integer, String> allThree = myJoinStoreThree.all();
+
+        List<KeyValue<Integer, String>> all = new LinkedList<>();
+
+        while (allOne.hasNext()) {
+            all.add(allOne.next());
+        }
+        while (allTwo.hasNext()) {
+            all.add(allTwo.next());
+        }
+        while (allThree.hasNext()) {
+            all.add(allThree.next());
+        }
+        allOne.close();
+        allTwo.close();
+        allThree.close();
+
+        for (KeyValue<Integer, String> elem : all) {
+            assertTrue(expectedResult.contains(elem));
+        }
+    }
+
+    private KafkaStreams prepareTopology(final String queryableName) {
+        final StreamsBuilder builder = new StreamsBuilder();
+
+        //TODO - bellemare - this appears to only work when I materialize the state... because I cannot access auto-materialized stateSupplier in internalStreamsBuilder!
```

Comment:
I'm required to supply a materialized store with a name, otherwise the internals of KTableImpl cannot get the name for the state store, and therefore the SubscriptionResolverJoinProcessorSupplier cannot access it (fails with NPE). This is related to the aforementioned changes to the KTableImpl constructor dropping the `isQueryable` boolean.

Reply:
@guozhangwang 
Hi Guozhang - I noticed that in this pull ( https://github.com/apache/kafka/commit/c0353d8ddce88bac6fc04f85dd40cb95b8ca5cf9 ) you made it such that the non-queryable store name is no longer obtainable in KTableImpl. I need access to the materializedInternal.storeName, but cannot get it anymore (that I know of). Do you have any suggestions of what I should do to get it? My only other options appear to be rewriting the KTableImpl constructors back to what they were before, but since you made this change I would like your input.

Basically, I am now being given materializerInternal.queryableStoreName, but I need materializerInternal.storeName for when the user does not specify the materialized store (which is frequently).

Please advise - Thanks

Reply:
@bellemare Sorry I've just seen this now. The main motivation of that PR is to avoid materializing state stores unnecessarily and the queryable-name refacotring is sort of a by-product. I was thinking that the parent store names (user-specified or internal) can be accessed from `KTableValueGetterSupplier#storeNames`, would that be a possible work-around for you?

Reply:
@guozhangwang - thanks for the reply. Basically `KTableA.doJoinOnForeignKey(....) ` needs to be able to access `KTableA's` own internal state store to compare the hash values for the workflow we established in KIP-213. When `KTableA` is not materialized (internally or by user-specified queryable-name), then I do not think we can query the parents directly since they are by definition different data stores. I do not see how they can slot in as a replacement for the following section:
https://github.com/apache/kafka/blob/0d6d92e4a137ac06f3f6eb28495a38714a5734e1/streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResolverJoinProcessorSupplier.java#L86-L89

Based on my understanding, it appears I NEED to have TableA materialized, otherwise I am not sure how to proceed with this ticket. This is basically where I am stuck.
 
I admit I don't know enough about this part of Kafka Streams to speak with certainty on this.

Reply:
I see.

I think the protocol should be that when a join operator requires to access its parents (the joining KTables, for example in this case) it will quire from its parent's `valueGetter`, which would chase up until it finds some processor with associated state stores, or it would go all the way up and require the source KTable to be materialized (see `KTableImpl#valueGetterSupplier` --- actually you can find the related materials I got from Kafka Summit this year, https://www.slideshare.net/GuozhangWang/performance-analysis-and-optimizations-for-kafka-streams-applications-145539130 around slide 56 :)

So back to your example, even if the source KTable is materialized without a name, calling a join later should cause its `KTableSource#materialize()` be triggered, where its internal `storeName` which is always not-null since it would be auto-generated as  `nameProvider.newStoreName(generatedStorePrefix)` if not specified, would be assigned to its exposable `queryableName`. And then this name can be accessed by its descent processors in the topology if necessary.

Hope this would help for you to figure it out.

Reply:
Note the logic `should cause its KTableSource#materialize() be triggered` may need to be added by you, i.e. the `SubscriptionResolverJoinProcessorSupplier` should require to get its parent's valueGetter, which eventually would trigger this function.

Reply:
@guozhangwang I think I solved it with your original recommendation, though by getting the valueGetterSupplier as you recommended and not the storeNames directly. Your follow up message is exactly what I did, so I am glad to have that confirmation. It's passing my tests now and appears to work with and without user-specified queryableStoreNames, so I believe this did the trick! Thanks for your help! 👍 

Now I just need to work on the remaining highlighted issues.

Reply:
Great!